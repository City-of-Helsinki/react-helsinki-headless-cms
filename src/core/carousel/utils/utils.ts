/**
 * @param arr - any array.
 * @param len - the array length.
 * @return chunks array of initial array (grouped elements).
 */
export function splitArrayIntoChunksOfLen(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  arr: Array<React.ReactElement<any>>,
  len: number,
) {
  const chunks = [];
  let i = 0;
  const n = arr.length;
  while (i < n) {
    chunks.push(arr.slice(i, (i += len)));
  }
  return chunks;
}

/**
 * @param index - item set index.
 * @param itemSetId - item set (element of the array).
 * @return an unique key based on input data.
 */
export const getItemSetKey = (index: number, itemSetId?: string) =>
  `itemSet-${itemSetId ?? Math.random()}-${index}`;

/**
 * Create an unique key for item set.
 * @param index - item index.
 * @param itemId - item (element of the array). Default 'itemSet-item'
 * @param itemSetPrefix - item prefix (element of the array).
 * @return an unique key based on input data.
 */
export const getItemSetItemKey = (
  index: number,
  itemId?: string,
  itemSetPrefix = 'itemSet-item',
) => `${itemSetPrefix}-${itemId ?? Math.random()}-${index}`;

/**
 * @param item - item (element of the array).
 * @param index - item index.
 * @return an unique key for the dot (slide counter) based on input data.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const getSlideDotKey = (item: any, index: number) =>
  `slide-dot-${item?.id ?? Math.random()}-${index}`;

export const getLoadMoreKey = () => `carousel-loadmore-${Math.random()}`;
