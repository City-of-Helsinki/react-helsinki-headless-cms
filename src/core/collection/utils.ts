import isPast from 'date-fns/isPast';

import type { EventType } from '../../common/eventsService/types';
import normalizeKeys from '../../linkedEvents/utils/normalizeKeys';
import { LINKEDEVENTS_DATE_NOW } from './constants';

/**
 * @param {ids} ids - the venue ids in number format.
 * @return {Array} - Returns array of tprek formatted string ids.
 */
export function getVenueIds(ids: number[]): string[] {
  return ids.map((id) => `tprek:${id}`);
}

/**
 * @param {event} - event.
 * @return {boolean} - Returns true if events' end time in past.
 */
export const isEventEndTimeInPast = (event: EventType): boolean =>
  !!event?.endTime && isPast(new Date(event.endTime));

/**
 * @param {event} - event.
 * @return {boolean} - Returns true if events' end time in null and start time in past.
 */
export const isEventEndTimeNullAndStartTimeInPast = (
  event: EventType,
): boolean =>
  !!event?.startTime &&
  event?.endTime === null &&
  isPast(new Date(event.startTime));

/**
 * @param {event} - event.
 * @return {boolean} - Returns true if event is closed.
 */
export const isEventClosed = (event: EventType): boolean =>
  isEventEndTimeInPast(event) || isEventEndTimeNullAndStartTimeInPast(event);

/**
 * @param {params} - Get params.
 * @return {object} - Returns normalized values.
 */
export const normalizeParamsValues = (params: Record<string, string>) => {
  const normalizedParams = { ...normalizeKeys(params) };

  // Fix for course event type lower case
  if (normalizedParams.eventType) {
    normalizedParams.eventType =
      normalizedParams.eventType.charAt(0).toUpperCase() +
      normalizedParams.eventType.slice(1);
  }

  // Fix boolean types
  Object.keys(normalizedParams).forEach((param) => {
    if (normalizedParams[param] === 'true') {
      normalizedParams[param] = true;
    } else if (normalizedParams[param] === 'false') {
      normalizedParams[param] = false;
    }
  });

  return normalizedParams;
};

/**
 * @param {dateString} - Get date string.
 * @return {boolean} - Returns true if the date is valid
 */
export function isValidDate(dateString: string) {
  return !Number.isNaN(Date.parse(dateString));
}

/**
 * @param {start} start - Get start date.
 * @return {boolean} - Returns current date as string if the initial was in past.
 */
export const convertDateStringInPastToNow = (start: string): string => {
  const startDate = new Date(start);
  const now = new Date();

  if (isValidDate(start) && startDate <= now) {
    return LINKEDEVENTS_DATE_NOW;
  }
  return start;
};
