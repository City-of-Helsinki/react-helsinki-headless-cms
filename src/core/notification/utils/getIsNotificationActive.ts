import isFuture from 'date-fns/isFuture';
import isPast from 'date-fns/isPast';
import isSameDay from 'date-fns/isSameDay';

import { NotificationType } from '../../../common/headlessService/types';

export default function isNotificationActive(
  notification?: NotificationType | null,
) {
  const { startDate, endDate, title, content } = notification ?? {};
  const isNotificationContentMissing = !content && !title;

  if (!notification || isNotificationContentMissing) {
    return false;
  }

  const isSameDayOrFuture = (date: Date) =>
    isSameDay(new Date(), date) || isFuture(date);

  return [
    // both start and end dates are defined
    () =>
      startDate &&
      endDate &&
      isPast(new Date(startDate)) &&
      isSameDayOrFuture(new Date(endDate)),
    // only end date defined
    () => !startDate && endDate && isSameDayOrFuture(new Date(endDate)),
    // only start date defined
    () => startDate && !endDate && isPast(new Date(startDate)),
    // neither start or end date are defined
    () => !startDate && !endDate,
  ].some((f) => !!f());
}
