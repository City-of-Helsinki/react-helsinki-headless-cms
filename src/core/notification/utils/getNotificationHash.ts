import hash from "../../../common/utils/hash";
import { NotificationType } from "../../../common/headlessService/types";

// The notification is a singleton in the CMS. Instead of there being multiple
// notifications, there ever only is one which is edited according to current
// needs.

// Because this notification comes with a dismissal option, we have to be able
// to distinct different versions of the notification, so that we can safely
// determine that the user has previously dismissed this notification (and
// received) the urgent information that's within the notification.
export default function getNotificationHash(
  notification?: NotificationType | null
) {
  // Notification fields used to build string for unique hash
  const stringKeys = [
    "title",
    "content",
    "linkText",
    "linkUrl",
    "level",
    "endDate",
    "startDate",
  ] as const;
  const combinedString = stringKeys.reduce(
    (acc, current) => acc + (notification?.[current] ?? ""),
    ""
  );

  return hash(combinedString);
}
