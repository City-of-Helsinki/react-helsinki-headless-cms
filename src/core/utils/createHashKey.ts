const MAX_KEY_LENGTH = 200;
export default function createHashKey(source: string | object) {
  const data = typeof source === 'string' ? source : JSON.stringify(source);
  return btoa(unescape(encodeURIComponent(data))).substring(0, MAX_KEY_LENGTH);
}
