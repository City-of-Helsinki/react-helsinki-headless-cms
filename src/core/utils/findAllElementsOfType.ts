import { getChildrenByType } from './getChildrenByType';
import { recursiveMap } from './recursiveMap';

/**
 * Find all elements of defined types recursively from a ReactNode.
 * @returns a list of elements of specified types that could be found from inside the ReactNode.
 */
export const findAllElementsOfType = (
  children: React.ReactNode,
  types: string[],
) => {
  const elementsOfType = [];
  recursiveMap(children, (child) => {
    elementsOfType.push(getChildrenByType(child, types));
    return child;
  });
  return elementsOfType.flat();
};
