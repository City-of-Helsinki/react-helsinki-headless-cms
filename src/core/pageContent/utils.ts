import { EventType } from '../../common/eventsService/types';
import {
  ArticleType,
  PageModule,
  PageType,
} from '../../common/headlessService/types';
import {
  getElementTextContent,
  isArticleType,
  isEventSearch,
  isEventSearchCarousel,
  isEventSelected,
  isEventSelectedCarousel,
  isEventType,
  isLayoutArticle,
  isLayoutArticleCarousel,
  isLayoutPage,
  isLayoutPageCarousel,
  isPageType,
} from '../../common/headlessService/utils';
import { CardProps } from '../card/Card';
import { type CollectionProps } from '../collection/Collection';
import {
  CollectionType,
  EventSearchCollectionType,
  EventSelectionCollectionType,
  GeneralCollectionType,
} from '../collection/types';

export function getCollections(
  pageModules: PageModule[],
  isEventSearchCollectionsEnables = false,
): CollectionType[] {
  return pageModules?.reduce((collections, module, index) => {
    const commonFields: CollectionType = {
      id: index.toString(),
      title: 'title' in module ? module.title : null,
      description: 'description' in module ? module.description : null,
      items: [],
      // eslint-disable-next-line no-underscore-dangle
      __typename: module.__typename,
    };

    if (isLayoutArticle(module) || isLayoutArticleCarousel(module)) {
      collections.push({
        ...commonFields,
        items: module.articles,
        showAllUrl: module.showAllLink,
      });
    }
    if (isLayoutPage(module) || isLayoutPageCarousel(module)) {
      collections.push({
        ...commonFields,
        items: module.pages,
      });
    }
    if (isEventSearchCollectionsEnables) {
      if (isEventSearch(module) || isEventSearchCarousel(module)) {
        collections.push({
          ...commonFields,
          url: module.url,
        } as EventSearchCollectionType);
      }
      if (isEventSelected(module) || isEventSelectedCarousel(module)) {
        collections.push({
          ...commonFields,
          events: module.events,
        } as EventSelectionCollectionType);
      }
    }
    return collections;
  }, []);
}

export function getArticlePageCardProps(
  item: ArticleType | PageType,
): CardProps {
  return {
    id: item.id,
    title: item.title,
    url: item.link,
    imageUrl: item.featuredImage?.node?.mediaItemUrl,
    ariaLabel: item.title,
    text: getElementTextContent((item.lead || item.content) ?? ''),
    hasLink: true,
    withBorder: true,
    withShadow: false,
    clampText: true,
    direction: 'responsive' as CardProps['direction'],
    openLinkInNewTab: false,
  };
}

export function getEventCardProps(item: EventType, locale = 'fi'): CardProps {
  const image = item.images.length > 0 ? item.images[0] : null;
  return {
    id: item.id,
    title: item.name[locale],
    url: item.internalId,
    imageUrl: image?.url,
    ariaLabel: item.name[locale],
    hasLink: true,
    withBorder: true,
    withShadow: false,
    clampText: true,
    direction: 'responsive' as CardProps['direction'],
    openLinkInNewTab: false,
  };
}

export function getCollectionCards(
  collection: GeneralCollectionType,
  locale = 'fi',
): CardProps[] {
  return collection.items.reduce((result: CardProps[], item) => {
    if (isPageType(item) || isArticleType(item))
      result.push(getArticlePageCardProps(item));
    else if (isEventType(item)) result.push(getEventCardProps(item, locale));

    return result;
  }, []);
}

export function getCollectionUIType(
  collection: CollectionType,
): CollectionProps['type'] {
  // eslint-disable-next-line no-underscore-dangle
  return collection.__typename.includes('Carousel') ? 'carousel' : 'grid';
}
