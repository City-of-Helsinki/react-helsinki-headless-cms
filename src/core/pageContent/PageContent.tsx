import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import classNames from 'classnames';
import { KorosType } from 'hds-react';

import { useConfig } from '../configProvider/useConfig';
import SidebarContent from './sidebarContent/SidebarContent';
import { PageContentLayout } from './PageContentLayout';
import { PageMainContent } from './PageMainContent';
import PageContentBreadcrumbs from './PageContentBreadcrumbs';
import { Breadcrumb, HeroProps } from './types';
import { PageMeta } from './meta/PageMeta';
import {
  Collection,
  PageArticleCollection,
  EventSearchCollection,
  EventSelectionCollection,
  LocationsSelectionCollection,
} from '../collection/Collection';
import {
  ArticleType,
  LanguageCodeEnum,
  PageType,
} from '../../common/headlessService/types';
import { getCollections, getCollectionUIType } from './utils';
import {
  isEventSearchCollection,
  isEventSelectionCollection,
  isLayoutCard,
  isLayoutCards,
  isLayoutContent,
  isLayoutImage,
  isLayoutSteps,
  isLocationsSelectionCollection,
  isPageType,
} from '../../common/headlessService/utils';
import { ContentModule } from '../pageModules/ContentModule/ContentModule';
import { CardModule } from '../pageModules/CardModule/CardModule';
import { CardsModule } from '../pageModules/CardsModule/CardsModule';
import { ImageModule } from '../pageModules/ImageModule/ImageModule';
import { StepsModule } from '../pageModules/StepsModule/StepsModule';
import createHashKey from '../utils/createHashKey';

export type PageContentProps = {
  page?: PageType | ArticleType;
  breadcrumbs?: Breadcrumb[];
  content?:
    | React.ReactNode
    | ((page: PageType | ArticleType) => React.ReactNode);
  shareLinks?: React.ReactNode;
  collections?:
    | React.ReactElement<typeof Collection>[]
    | ((
        page: PageType | ArticleType,
      ) => React.ReactElement<typeof Collection>[]);
  heroContainer?: JSX.Element;
  backUrl?: string;
  sidebarContentProps?: Partial<typeof SidebarContent>;
  PageContentLayoutComponent?: typeof PageContentLayout;
  className?: string;
  onArticlesSearch?: (tag: string) => void;
  // All other props
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [x: string]: any;
} & Partial<typeof PageContentLayout>;

// Modules: Content, Image, Cards, Steps (possibly other in future)
export const defaultContentModules = (
  page: PageType | ArticleType,
): React.ReactNode[] => {
  const contentModules: React.ReactNode[] = [];
  page?.modules?.map((module, index) => {
    const uniqueKey = createHashKey(`${index}-${JSON.stringify(module)}`);
    if (isLayoutContent(module)) {
      contentModules.push(
        <ContentModule
          key={uniqueKey}
          content={module.content}
          backgroundColor={module.backgroundColor}
        />,
      );
    } else if (isLayoutCard(module)) {
      contentModules.push(
        <CardModule
          key={uniqueKey}
          title={module.title}
          text={module.description}
          backgroundColor={module.backgroundColor}
          hasLink
          url={module.link.url}
          imageUrl={module.image.medium_large}
          imagePosition={
            module.alignment?.indexOf('left') === -1
              ? 'image-right'
              : 'image-left'
          }
          isDelimited={module.alignment.startsWith('delimited')}
        />,
      );
    } else if (isLayoutCards(module)) {
      contentModules.push(<CardsModule key={uniqueKey} items={module.cards} />);
    } else if (isLayoutImage(module)) {
      contentModules.push(<ImageModule key={uniqueKey} />);
    } else if (isLayoutSteps(module)) {
      contentModules.push(
        <StepsModule
          key={uniqueKey}
          title={module.title}
          steps={module.steps.map((step) => ({
            title: step.title,
            content: step.content,
          }))}
          helpText={module.description}
          color={module.color}
          type={module.type}
        />,
      );
    }
    return null;
  });
  return contentModules;
};

export const defaultContent = (
  page: PageType | ArticleType,
  onArticlesSearch?: (tag: string) => void,
) => {
  let hideTitle = false;
  if (isPageType(page)) {
    hideTitle = Boolean(page?.hero?.title);
  }

  return (
    <PageMainContent
      title={!hideTitle && page.title}
      content={page?.content}
      date={(page as ArticleType)?.date}
      categories={(page as ArticleType)?.categories}
      contentModules={defaultContentModules(page)}
      onArticlesSearch={onArticlesSearch}
    />
  );
};

export const defaultCollections = ({
  page,
  isEventModulesEnabled = true,
  isVenueModulesEnabled = true,
}: {
  page: PageType | ArticleType;
  isEventModulesEnabled: boolean;
  isVenueModulesEnabled: boolean;
}) =>
  getCollections(page?.modules, true)?.reduce(
    (collectionElements, collection) => {
      const commonCollectionProps = {
        key: `collection-${Math.random()}`,
        title: collection.title,
        description: collection.description,
        type: getCollectionUIType(collection),
        collectionContainerProps: { withDots: false },
      };

      if (isEventSearchCollection(collection)) {
        if (isEventModulesEnabled) {
          collectionElements.push(
            <EventSearchCollection
              {...commonCollectionProps}
              collection={collection}
            />,
          );
        }
      } else if (isEventSelectionCollection(collection)) {
        if (isEventModulesEnabled) {
          collectionElements.push(
            <EventSelectionCollection
              {...commonCollectionProps}
              collection={collection}
            />,
          );
        }
      } else if (isLocationsSelectionCollection(collection)) {
        if (isVenueModulesEnabled) {
          collectionElements.push(
            <LocationsSelectionCollection
              {...commonCollectionProps}
              collection={collection}
              locale={page.language.locale as LanguageCodeEnum}
            />,
          );
        }
      } else {
        collectionElements.push(
          <PageArticleCollection
            {...commonCollectionProps}
            collection={collection}
          />,
        );
      }
      return collectionElements;
    },
    [],
  );

export function PageContent(props: PageContentProps) {
  const {
    page,
    breadcrumbs,
    collections,
    heroContainer,
    backUrl,
    sidebarContentProps,
    PageContentLayoutComponent = PageContentLayout,
    content,
    shareLinks,
    className,
    onArticlesSearch,
    ...pageContentLayoutProps
  } = props;

  const {
    components: { Head },
    eventsApolloClient,
    venuesApolloClient,
    mainContentId,
  } = useConfig();

  const isEventModulesEnabled =
    eventsApolloClient !== undefined && eventsApolloClient !== 'disabled';
  const isVenueModulesEnabled =
    eventsApolloClient !== undefined && venuesApolloClient !== 'disabled';

  const getHeroProps = () => {
    const heroProps: HeroProps = {};
    if (isPageType(page)) {
      heroProps.title = page.hero?.title || '';
      heroProps.description = page.hero?.description;
      heroProps.backgroundColor = page.hero?.background_color;
      heroProps.korosType = (page.hero?.wave_motif as KorosType) || 'basic';
      heroProps.actionUrl = page.hero?.link.url;
      heroProps.actionUrlTarget = page.hero?.link.target;
      heroProps.actionText = page.hero?.link?.title;
      heroProps.isPageType = true;
    }
    return heroProps;
  };

  return (
    <main
      id={mainContentId || 'main-content'}
      className={classNames('page-main-content', className)}
    >
      {Head && <PageMeta headComponent={Head} page={page} />}
      <PageContentLayoutComponent
        {...props}
        {...pageContentLayoutProps}
        {...getHeroProps()}
        breadcrumbs={
          breadcrumbs && <PageContentBreadcrumbs breadcrumbs={breadcrumbs} />
        }
        heroContainer={heroContainer}
        id={page?.id ?? 'page'}
        imageSrc={page?.featuredImage?.node?.mediaItemUrl}
        imageAlt={page?.featuredImage?.node?.altText}
        imageLabel={page?.featuredImage?.node?.photographerName}
        backUrl={backUrl}
        content={
          typeof content === 'function'
            ? content(page)
            : content ?? defaultContent(page, onArticlesSearch)
        }
        shareLinks={shareLinks}
        collections={
          typeof collections === 'function'
            ? collections(page)
            : collections ??
              defaultCollections({
                page,
                isEventModulesEnabled,
                isVenueModulesEnabled,
              })
        }
        sidebarContent={
          <SidebarContent
            content={(page as PageType)?.sidebar}
            {...sidebarContentProps}
          />
        }
      />
    </main>
  );
}
