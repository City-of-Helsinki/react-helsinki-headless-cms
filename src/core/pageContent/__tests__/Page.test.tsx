import React from 'react';

import { render, screen } from '../../../common/utils/testingLibrary';
import pageMock, { sidebarLinkList } from '../__mocks__/page.mock';
import { PageContent } from '../PageContent';

test('renders page with expected content', () => {
  render(<PageContent page={pageMock} />);

  // Title of page is presented correctly
  expect(
    screen.getByRole('heading', { name: pageMock.title, level: 1 }),
  ).toBeInTheDocument();
  // Featured image is included on the page
  expect(
    screen.getByRole('img', {
      name: pageMock.featuredImage.node.altText,
    }),
  ).toBeInTheDocument();
  // Check content
  expect(screen.getByRole('article')).toMatchSnapshot();

  // Sidebar contains expected links
  expect(
    screen.getByRole('heading', { name: sidebarLinkList.title, level: 2 }),
  ).toBeInTheDocument();
  expect(screen.getByText(sidebarLinkList.description)).toBeInTheDocument();
  expect(document.getElementById(sidebarLinkList.anchor)).toBeInTheDocument();
  sidebarLinkList.links.forEach((link) => {
    const linkElement = screen.getByRole('link', {
      name: new RegExp(link.title, 'g'),
    });

    expect(linkElement).toBeInTheDocument();
    expect(linkElement).toHaveAttribute('href', link.url);

    if (link.target === '_blank') {
      expect(linkElement).toHaveAttribute('target', '_blank');
      expect(linkElement).toHaveAttribute('rel', 'noopener noreferrer');
      expect(linkElement).toHaveAttribute(
        'aria-label',
        expect.stringContaining('opens in a new tab'),
      );
    }
  });
});
