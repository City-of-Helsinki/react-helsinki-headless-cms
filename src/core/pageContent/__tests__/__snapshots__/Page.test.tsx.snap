// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders page with expected content 1`] = `
<article
  class="mainContent"
>
  <header>
    <h1
      class="text h2 "
    >
      Kulttuurikasvatus
    </h1>
  </header>
  
  
  <p
    class="text body "
  >
    <strong>
      Taiteen ja kulttuurin Helsinki
    </strong>
  </p>
  
  
  <p
    class="text body "
  >
    Taiteen ja kulttuurin näkökulmasta helsinkiläiset lapset ja nuoret ovat poikkeuksellisessa asemassa. Kaupungissamme on valtavasti uusia ja yllätyksellisiä ympäristöjä aistia, kokea, oppia uutta ja osallistua maailmanmenoon. Kaupungissa toimii laaja joukko taiteen ja kulttuuriperinnön toimijoita, joista lukuisat tarjoavat sisältöjä lapsille ja nuorille, myös varhaiskasvatus- tai koulupäivän yhteyteen. Monipuolisten kokemusten äärelle pääsee kuitenkin vain osa lapsista ja nuorista. Taustalla ovat mm. huoltajien sosioekonominen tilanne sekä perheen kiinnostus ja tieto taide- ja kulttuurisisällöistä. Kulttuurikasvatusta ja kulttuurikasvatussuunnitelmaa tarvitaan, jotta taide ja kulttuuri olisivat tasapuolisesti osa jokaisen lapsen ja nuoren arkea ja elämänikänen yhteys taiteeseen ja kulttuuriin saa hyvän pohjan.
  </p>
  
  
  <p
    class="text body "
  >
    <strong>
      Lapsi ja nuori keskiössä
    </strong>
  </p>
  
  
  <p
    class="text body "
  >
    Jokainen lapsi on syntyessään aktiivinen ympäristönsä tutkija. Lapsuus on tässä ja nyt. Lapsi tai nuori on aktiivinen toimija, kokija ja näkijä. Lapsuus ja nuoruus ovat ihmisen elinkaaressa itsessään ainutlaatuisia, merkittäviä ja herkkiä ajanjaksoja. Lapsuuden aika vaikuttaa koko elinkaaren ajan, nuoruuteen, aikuisuuteen ja vanhuuteen asti. Lapsen ja nuoren kasvu kohti vahvaa, ilmaisevaa ja omaa elämäänsä hallitsevaa tulevaisuutta edellyttää turvallista, monipuolista ja toimijuutta vahvistavaa kasvuympäristöä.
  </p>
  
  
  <p
    class="text body "
  >
    <strong>
      Kulttuurikasvatussuunnitelma 
    </strong>
  </p>
  
  
  <p
    class="text body "
  >
    Kulttuurikasvatussuunnitelma Helsingissä toimii Kulttuuri- ja vapaa-aika -toimialan, Kasvatus ja koulutus -toimialan (sis. yksityisen koulut) sekä taide- ja kulttuuritoimijoiden välisenä ohjaavana suunnitelmana. Yhteistyön tavoitteena on säännöllisten ja helposti lähestyttävien toimintatapojen, kuten kulttuuripolkujen, viestintäpalveluiden (esim. Kultus.fi) ja alan verkostojen kehittäminen ja ylläpitäminen. Suunnitelma koskee varhaiskasvatusta ja kouluja. Taide- ja taitoaineiden tavoitteet määritellään opetussuunnitelmissa ja varhaiskasvatussuunnitelmissa. Kulttuurikasvatusta voi sisältyä myös näiden kokonaisuuksien sisään. 
  </p>
  
  
  <p
    class="text body "
  >
    Helsingin kulttuurikasvatussuunnitelma koskee kulttuuriperinnön, kulttuuriperintökasvatuksen, taiteen ja taidekasvatuksen sisältöjä, menetelmiä ja paikkoja. Se ei kuitenkaan ohjaa lapsen, nuoren tai perheen vapaa-ajan kulttuuri- ja taidekasvatuksen yhteistyön muotoja.
  </p>
  
  
</article>
`;
